ddtrace/bootstrap/sitecustomize.py

```diff
@@ -74,6 +69,88 @@ def update_patched_modules():
         EXTRA_PATCHED_MODULES[module] = asbool(should_patch)
 
 
+if PY2:
+    _unloaded_modules = []
+
+
+def gevent_is_installed():
+    # https://stackoverflow.com/a/51491863/735204
+    if sys.version_info >= (3, 4):
+        return importlib.util.find_spec("gevent")
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader("gevent")
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module("gevent")
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module("gevent")
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="auto").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        log.debug("skipping sitecustomize module unload because of DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == 0")
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        log.debug(
+            "skipping sitecustomize module unload because of invalid envvar value"
+            "DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == {}",
+            dd_unload_sitecustomize_modules,
+        )
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not gevent_is_installed():
+        log.debug(
+            "skipping sitecustomize module unload because DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == auto and "
+            "gevent is not installed"
+        )
+        return False
+    return True
+
+
+def cleanup_loaded_modules():
+    if not should_cleanup_loaded_modules():
+        return
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for m in list(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP):
+        if m.startswith("atexit"):
+            continue
+        if m.startswith("asyncio"):
+            continue
+        if m.startswith("attr"):
+            continue
+        if m.startswith("concurrent"):
+            continue
+        if m.startswith("ddtrace"):
+            continue
+        if m.startswith("logging"):
+            continue
+        if m.startswith("typing"):  # required by Python < 3.7
+            continue
```

Comment:
should we have a configuration for this?

I wonder how we should react to other modules that cause issues?

Reply:
should this be an explicit allow list rather than a deny list?

do we only need to do this on 1-2 stdlib modules only to resolve any issues?

Reply:
@brettlangdon I debated this with @Yun-Kim and @P403n1x87 and we collectively [decided](https://github.com/DataDog/dd-trace-py/pull/4863#discussion_r1098926561) that keeping an "allow list" in sync with changes to gevent would incur a bigger maintenance burden than maintaining this "deny" list. The way this change accounts for the open-endedness of possible undefined behavior stemming from that approach is by putting the module cloning functionality behind an off-by-default feature flag.

Reply:
From our offline discussion, we agreed to potentially make this more configurable in case we see the benefit in future releases.